proposals/nnnn-screenmanager-push-pop.md

```diff
@@ -0,0 +1,129 @@
+ */
+- (void)templateFailedToDeactivateWithError:(NSError *)error;
+
+@end
+```
+
+Developers may build classes that implement this protocol in order to separate their SDL code so that each "screen" gets its own class instead of all of their SDL code being all in one class. The template lifecycle methods are called by the screen manager as it sends RPCs to present and dismiss the templates displaying on the screen. It will use the `layout` property to determine which layout to present. If the next template layout to be shown is equivalent to the current template layout, then the RPC call will be skipped, but the lifecycle methods will still be called.
+
+### Screen Manager Additions
+
+Additions will be made to the screen manager:
+
+```objc
+@interface SDLScreenManager : NSObject
+
+// Everything that already exists
+
+/**
+ The current stack of templates.
+ */
+@property (strong, nonatomic, readonly) NSArray<id<SDLTemplate>> *templateStack;
+
+/**
+ The current template will deactivate and the new template that is passed will activate and be pushed onto the templateStack.
+ */
+- (void)pushTemplate:(id<SDLTemplate>)template;
+
+/**
+ The current template will deactivate and the previously displayed template (lastObject - 1 in the templateStack) will re-appear while the current template is removed from the templateStack. If there is only one object in the templateStack, nothing will happen.
+ */
+- (void)popTemplate;
+
+/**
+ The current template will deactivate and the template stack will be cleared except for the first item, which will either activate or be replaced if a new template is passed. That new template will then activate. If there is only one object in the templateStack and no newRootTemplate is passed, nothing will happen.
+ */
+- (void)popToRootTemplate:(nullable id<SDLTemplate>)newRootTemplate;
+
+/**
+ The current template will deactivate. Then the new template will activate by replacing the current template on the stack instead of pushing it onto the stack.
+ */
+- (void)replaceCurrentTemplateWithNewTemplate:(id<SDLTemplate>)newTemplate;
+
+/**
+ All text, images, and soft buttons will be cleared from the screen. Preloaded choices and main menu items will remain.
+ */
+- (void)clearScreen;
+
+@end
+```
+
+### Additional Implementation Notes
+1. When connected to systems running RPC <6.0, the screen manager should send `SetDisplayLayout` to change the template, and on systems running RPC >6.0, the screen manager should send `Show` to change the template.
+
+### Java APIs
+This will be set up in a similar way to the Obj-C APIs above. The `SDLTemplate` protocol and screen manager additions above would be translated into appropriate Java APIs. Any necessary changes are at the discretion of the project maintainers. However larger changes that would impact back into the Objective-C code above (such as adding or removing a method) will require proposal revisions.
```

Comment:
```suggestion
This will be set up in a similar way to the Obj-C APIs above. The `SDLTemplate` protocol and screen manager additions above would be translated into appropriate Java APIs. Any necessary changes are at the discretion of the Project Maintainer. However larger changes that would impact the Objective-C code above (such as adding or removing a method) will require proposal revisions.
```